crates/trigger-http/src/wasip3.rs

```diff
@@ -0,0 +1,109 @@
+use crate::{headers, server::HttpExecutor, TriggerInstanceBuilder};
+use anyhow::{Context, Result};
+use futures::{channel::oneshot, FutureExt};
+use http::{HeaderName, HeaderValue};
+use http_body_util::BodyExt;
+use spin_factors::RuntimeFactors;
+use spin_factors_executor::InstanceState;
+use spin_http::routes::RouteMatch;
+use std::net::SocketAddr;
+use tokio::task;
+use tracing::{instrument, Level};
+use wasmtime_wasi_http::{
+    body::HyperIncomingBody as Body,
+    p3::{
+        bindings::{http::types, ProxyIndices},
+        WasiHttpCtxView,
+    },
+};
+
+/// An [`HttpExecutor`] that uses the `wasi:http@0.3.*/handler` interface.
+pub(super) struct Wasip3HttpExecutor<'a>(pub(super) &'a ProxyIndices);
+
+impl HttpExecutor for Wasip3HttpExecutor<'_> {
+    #[instrument(name = "spin_trigger_http.execute_wasm", skip_all, err(level = Level::INFO), fields(otel.name = format!("execute_wasm_component {}", route_match.lookup_key().to_string())))]
+    async fn execute<F: RuntimeFactors>(
+        &self,
+        instance_builder: TriggerInstanceBuilder<'_, F>,
+        route_match: &RouteMatch<'_, '_>,
+        mut req: http::Request<Body>,
+        client_addr: SocketAddr,
+    ) -> Result<http::Response<Body>> {
+        let spin_http::routes::TriggerLookupKey::Component(component_id) = route_match.lookup_key()
+        else {
+            anyhow::bail!("INCONCEIVABLE");
+        };
+
+        tracing::trace!("Executing request using the Wasi executor for component {component_id}");
+
+        let (instance, mut store) = instance_builder.instantiate(()).await?;
+
+        let headers = headers::prepare_request_headers(&req, route_match, client_addr)?;
+        req.headers_mut().clear();
+        req.headers_mut()
+            .extend(headers.into_iter().filter_map(|(n, v)| {
+                let Ok(name) = n.parse::<HeaderName>() else {
+                    return None;
+                };
+                let Ok(value) = HeaderValue::from_bytes(v.as_bytes()) else {
+                    return None;
+                };
+                Some((name, value))
+            }));
+
+        let getter = (|data| wasi_http::<F>(data).unwrap())
+            as fn(&mut InstanceState<F::InstanceState, ()>) -> WasiHttpCtxView<'_>;
+
+        let (request, body) = req.into_parts();
+        let body = body.map_err(spin_factor_outbound_http::p2_to_p3_error_code);
+        let request = http::Request::from_parts(request, body);
+        let (request, request_io_result) = types::Request::from_http(request);
+        let request = wasi_http::<F>(store.data_mut())?.table.push(request)?;
+
+        let guest = self.0.load(&mut store, &instance)?;
+
+        let (tx, rx) = oneshot::channel();
+        task::spawn(
+            async move {
```

Comment:
I'm guessing I need to sprinkle some `.in_current_span()` calls here, but not sure where.  Do we need them on every `async` block and every call to an async function?

Reply:
You typically only need it for the outermost `async` block within a `spawn`. If you peel back several layers of indirection it is essentially pulling the "current span" (handle) from a thread-local and then wrapping this `async`'s Future in an [`Instrumented`](https://docs.rs/tracing/0.1.41/tracing/instrument/struct.Instrumented.html) that sets that thread-local every time it is polled.